File: nodejs/src/client/connections.ts
/**
 * Connections API. create, update, test, list, get.
 * Backend: workflows microservice /workflows/connections.
 */

import type { LoxtepHttpClient } from '../http/client.js';
import type {
  Connection,
  ConnectionCreateInput,
  ConnectionUpdateInput,
  ConnectionsListFilters,
  ConnectionsListResponse,
  ConnectionTestResult,
} from './connection-types.js';

const CONNECTIONS_BASE = '/workflows/connections';

function buildQueryString(
  params: Record<string, string | number | boolean | string[] | undefined>
): string {
  const search = new URLSearchParams();
  for (const [k, v] of Object.entries(params)) {
    if (v === undefined || v === null) continue;
    if (Array.isArray(v)) {
      search.set(k, v.join(','));
    } else {
      search.set(k, String(v));
    }
  }
  const qs = search.toString();
  return qs ? `?${qs}` : '';
}

/**
 * Create the connections API surface (get, list, create, update, test).
 */
export function createConnectionsApi(http: LoxtepHttpClient): {
  get: (id: string) => Promise<Connection>;
  list: (filters?: ConnectionsListFilters) => Promise<ConnectionsListResponse['data']>;
  create: (config: ConnectionCreateInput) => Promise<Connection>;
  update: (id: string, config: ConnectionUpdateInput) => Promise<Connection>;
  delete: (id: string) => Promise<void>;
  test: (id: string) => Promise<ConnectionTestResult>;
} {
  return {
    async get(id: string): Promise<Connection> {
      const res = await http.get<{ success: true; data: Connection }>(
        `${CONNECTIONS_BASE}/${encodeURIComponent(id)}`
      );
      return res.data;
    },

    async list(filters?: ConnectionsListFilters): Promise<ConnectionsListResponse['data']> {
      const params: Record<string, string | number | boolean | string[] | undefined> = {
        page: filters?.page ?? 1,
        page_size: filters?.page_size ?? 50,
      };
      if (filters?.search) params.search = filters.search;
      if (filters?.type) params.type = Array.isArray(filters.type) ? filters.type : [filters.type];
      if (filters?.status)
        params.status = Array.isArray(filters.status) ? filters.status : [filters.status];
      if (filters?.verified !== undefined) params.verified = filters.verified;
      if (filters?.draft !== undefined) params.draft = filters.draft;
      const qs = buildQueryString(params);
      const res = await http.get<ConnectionsListResponse>(`${CONNECTIONS_BASE}${qs}`);
      return res.data;
    },

    async create(config: ConnectionCreateInput): Promise<Connection> {
      const res = await http.post<{ success: true; data: Connection }>(CONNECTIONS_BASE, config);
      return res.data;
    },

    async update(id: string, config: ConnectionUpdateInput): Promise<Connection> {
      const res = await http.put<{ success: true; data: Connection }>(
        `${CONNECTIONS_BASE}/${encodeURIComponent(id)}`,
        config
      );
      return res.data;
    },

    async delete(id: string): Promise<void> {
      await http.delete(`${CONNECTIONS_BASE}/${encodeURIComponent(id)}`);
    },

    async test(id: string): Promise<ConnectionTestResult> {
      const res = await http.post<{ success: true; data: ConnectionTestResult }>(
        `${CONNECTIONS_BASE}/${encodeURIComponent(id)}/test`,
        {}
      );
      return res.data;
    },
  };
}

File: python/src/loxtep/connections.py
"""
Connections API. list, get, create, update, test.
"""

from typing import Any, Optional

from .http_client import AsyncLoxtepHttpClient, LoxtepHttpClient

CONNECTIONS_BASE = "/workflows/connections"


def _query_string(params: dict[str, Any]) -> str:
    parts = [f"{k}={v}" for k, v in params.items() if v is not None]
    return "?" + "&".join(parts) if parts else ""


class ConnectionsApi:
    """Sync connections surface."""

    def __init__(self, http: LoxtepHttpClient) -> None:
        self._http = http

    def get(self, id: str) -> dict[str, Any]:
        res = self._http.get(f"{CONNECTIONS_BASE}/{id}")
        return res.get("data", res) if isinstance(res, dict) else res

    def list(
        self,
        *,
        page: int = 1,
        page_size: int = 50,
        search: Optional[str] = None,
        type: Optional[str] = None,
        status: Optional[str] = None,
    ) -> dict[str, Any]:
        params: dict[str, Any] = {"page": page, "page_size": page_size}
        if search is not None:
            params["search"] = search
        if type is not None:
            params["type"] = type
        if status is not None:
            params["status"] = status
        qs = _query_string(params)
        res = self._http.get(f"{CONNECTIONS_BASE}{qs}")
        return res.get("data", res) if isinstance(res, dict) else res

    def create(
        self,
        name: str,
        type: str,
        key: str,
        *,
        data: Optional[str] = None,
        configuration: Optional[dict[str, Any]] = None,
        metadata: Optional[dict[str, Any]] = None,
    ) -> dict[str, Any]:
        body: dict[str, Any] = {"name": name, "type": type, "key": key}
        if data is not None:
            body["data"] = data
        if configuration is not None:
            body["configuration"] = configuration
        if metadata is not None:
            body["metadata"] = metadata
        res = self._http.post(CONNECTIONS_BASE, body)
        return res.get("data", res) if isinstance(res, dict) else res

    def update(
        self,
        id: str,
        *,
        name: Optional[str] = None,
        data: Optional[str] = None,
        configuration: Optional[dict[str, Any]] = None,
        metadata: Optional[dict[str, Any]] = None,
    ) -> dict[str, Any]:
        body: dict[str, Any] = {}
        if name is not None:
            body["name"] = name
        if data is not None:
            body["data"] = data
        if configuration is not None:
            body["configuration"] = configuration
        if metadata is not None:
            body["metadata"] = metadata
        res = self._http.put(f"{CONNECTIONS_BASE}/{id}", body)
        return res.get("data", res) if isinstance(res, dict) else res

    def delete(self, id: str) -> None:
        self._http.delete(f"{CONNECTIONS_BASE}/{id}")

    def test(self, id: str) -> dict[str, Any]:
        res = self._http.post(f"{CONNECTIONS_BASE}/{id}/test", {})
        return res.get("data", res) if isinstance(res, dict) else res


class AsyncConnectionsApi:
    """Async connections surface."""

    def __init__(self, http: AsyncLoxtepHttpClient) -> None:
        self._http = http

    async def get(self, id: str) -> dict[str, Any]:
        res = await self._http.get(f"{CONNECTIONS_BASE}/{id}")
        return res.get("data", res) if isinstance(res, dict) else res

    async def list(
        self,
        *,
        page: int = 1,
        page_size: int = 50,
        search: Optional[str] = None,
        type: Optional[str] = None,
        status: Optional[str] = None,
    ) -> dict[str, Any]:
        params: dict[str, Any] = {"page": page, "page_size": page_size}
        if search is not None:
            params["search"] = search
        if type is not None:
            params["type"] = type
        if status is not None:
            params["status"] = status
        qs = _query_string(params)
        res = await self._http.get(f"{CONNECTIONS_BASE}{qs}")
        return res.get("data", res) if isinstance(res, dict) else res

    async def create(
        self,
        name: str,
        type: str,
        key: str,
        *,
        data: Optional[str] = None,
        configuration: Optional[dict[str, Any]] = None,
        metadata: Optional[dict[str, Any]] = None,
    ) -> dict[str, Any]:
        body: dict[str, Any] = {"name": name, "type": type, "key": key}
        if data is not None:
            body["data"] = data
        if configuration is not None:
            body["configuration"] = configuration
        if metadata is not None:
            body["metadata"] = metadata
        res = await self._http.post(CONNECTIONS_BASE, body)
        return res.get("data", res) if isinstance(res, dict) else res

    async def update(
        self,
        id: str,
        *,
        name: Optional[str] = None,
        data: Optional[str] = None,
        configuration: Optional[dict[str, Any]] = None,
        metadata: Optional[dict[str, Any]] = None,
    ) -> dict[str, Any]:
        body: dict[str, Any] = {}
        if name is not None:
            body["name"] = name
        if data is not None:
            body["data"] = data
        if configuration is not None:
            body["configuration"] = configuration
        if metadata is not None:
            body["metadata"] = metadata
        res = await self._http.put(f"{CONNECTIONS_BASE}/{id}", body)
        return res.get("data", res) if isinstance(res, dict) else res

    async def delete(self, id: str) -> None:
        await self._http.delete(f"{CONNECTIONS_BASE}/{id}")

    async def test(self, id: str) -> dict[str, Any]:
        res = await self._http.post(f"{CONNECTIONS_BASE}/{id}/test", {})
        return res.get("data", res) if isinstance(res, dict) else res
